File: VideoMergeAndroid/app/src/main/java/com/moon/videomerger/editor/engine/TextWatermarkRenderer.kt
package com.moon.videomerger.editor.engine

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import java.io.File
import java.io.FileOutputStream

/**
 * 文字水印工具 —— 用 Android Canvas 生成透明背景的文字 PNG，供 FFmpeg overlay 使用。
 *
 * 对应 Python 脚本：text_watermark.py（用 Pillow 生成 PNG + ffmpeg overlay）。
 * FFmpegKit min 版本不含 drawtext 滤镜，所以采用相同的 PNG overlay 方案。
 */
object TextWatermarkRenderer {

    /**
     * 生成文字水印 PNG 文件。
     *
     * @param text 文字内容
     * @param fontSize 字号（px）
     * @param colorStr 颜色（white/red/yellow/cyan/#RRGGBB）
     * @param opacity 透明度 0~1
     * @param border 是否描边（黑色，增强可读性）
     * @return PNG 文件，和实际文字图像的宽高
     */
    fun renderToPng(
        context: Context,
        text: String,
        fontSize: Int,
        colorStr: String,
        opacity: Float,
        border: Boolean
    ): TextPngResult {
        val textColor = parseColor(colorStr)
        val alpha = (255 * opacity.coerceIn(0f, 1f)).toInt()

        // 测量文字尺寸
        val measurePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            textSize = fontSize.toFloat()
            typeface = Typeface.DEFAULT
        }
        val textBounds = Rect()
        measurePaint.getTextBounds(text, 0, text.length, textBounds)
        val textW = textBounds.width()
        val textH = textBounds.height()

        // 描边余量
        val pad = if (border) (fontSize / 8).coerceAtLeast(2) else 4
        val imgW = textW + pad * 4
        val imgH = textH + pad * 4

        // 创建透明 Bitmap
        val bitmap = Bitmap.createBitmap(imgW, imgH, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)

        val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            textSize = fontSize.toFloat()
            typeface = Typeface.DEFAULT
            color = applyAlpha(textColor, alpha)
            textAlign = Paint.Align.LEFT
        }

        // 文字基线位置（让文字居中）
        val baseline = pad * 2 + textH - textBounds.bottom.toFloat()

        // 绘制描边（黑色，多次偏移绘制模拟描边）
        if (border) {
            val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
                textSize = fontSize.toFloat()
                typeface = Typeface.DEFAULT
                color = applyAlpha(Color.BLACK, alpha)
                textAlign = Paint.Align.LEFT
                style = Paint.Style.FILL_AND_STROKE
                strokeWidth = (fontSize / 6).toFloat().coerceAtLeast(2f)
            }
            canvas.drawText(text, pad * 2 - textBounds.left.toFloat(), baseline, strokePaint)
        }

        // 绘制文字
        canvas.drawText(text, pad * 2 - textBounds.left.toFloat(), baseline, textPaint)

        // 保存为 PNG
        val pngFile = File(context.cacheDir, "text_watermark_${System.currentTimeMillis()}.png")
        FileOutputStream(pngFile).use { out ->
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
        }
        bitmap.recycle()

        return TextPngResult(
            file = pngFile,
            width = imgW,
            height = imgH
        )
    }

    /** 计算文字在视频上的 overlay 位置（9 宫格） */
    fun calcOverlayPosition(
        posKey: String,
        textW: Int,
        textH: Int,
        videoW: Int,
        videoH: Int,
        margin: Int = 20
    ): Pair<Int, Int> {
        return when (posKey) {
            "top-left" -> margin to margin
            "top-right" -> (videoW - textW - margin) to margin
            "bottom-left" -> margin to (videoH - textH - margin)
            "bottom-right" -> (videoW - textW - margin) to (videoH - textH - margin)
            "center" -> ((videoW - textW) / 2) to ((videoH - textH) / 2)
            else -> (videoW - textW - margin) to (videoH - textH - margin)
        }
    }

    private fun parseColor(str: String): Int {
        return when (str.lowercase()) {
            "white" -> Color.WHITE
            "black" -> Color.BLACK
            "red" -> Color.RED
            "green" -> Color.GREEN
            "blue" -> Color.BLUE
            "yellow" -> Color.YELLOW
            "cyan" -> Color.CYAN
            "magenta" -> Color.MAGENTA
            "gray", "grey" -> Color.GRAY
            else -> {
                if (str.startsWith("#") && str.length == 7) {
                    try {
                        Color.parseColor(str)
                    } catch (_: Exception) { Color.WHITE }
                } else {
                    Color.WHITE
                }
            }
        }
    }

    private fun applyAlpha(color: Int, alpha: Int): Int {
        return (alpha shl 24) or (color and 0x00FFFFFF)
    }
}

/** 文字 PNG 生成结果 */
data class TextPngResult(
    val file: File,
    val width: Int,
    val height: Int
)

File: VideoMergeIOS/VideoMerger/Editor/Engine/TextWatermarkRenderer.swift
//
//  TextWatermarkRenderer.swift
//  VideoMerger
//
//  iOS 移植自 Android TextWatermarkRenderer.kt
//  Canvas 渲染文字为透明 PNG（ffmpeg GPL 版才有 drawtext，统一用 PNG overlay 方案）
//

import UIKit

enum TextWatermarkRenderer {

    /// 渲染文字为 PNG 文件（带黑色描边增强可读性）
    static func renderToPng(text: String, fontSize: Int, colorStr: String,
                            opacity: Double, border: Bool) -> TextPngResult? {
        let color = parseColor(colorStr)
        // ★ 常规字重（对齐 Android Typeface.DEFAULT，原 bold 会比 Android 明显偏粗）
        let font = UIFont.systemFont(ofSize: CGFloat(max(fontSize, 12)))
        // 描边偏移随字号缩放（对齐 Android strokeWidth = fontSize/6，最小 2px）
        let strokeOffset = max(CGFloat(fontSize) / 6, 2)
        let attrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: color.withAlphaComponent(CGFloat(opacity)),
        ]

        let size = (text as NSString).size(withAttributes: attrs)
        let pad: CGFloat = border ? 10 : 2
        let width = Int(ceil(size.width)) + Int(pad) * 2
        let height = Int(ceil(size.height)) + Int(pad) * 2
        guard width > 0, height > 0 else { return nil }

        let renderer = UIGraphicsImageRenderer(size: CGSize(width: width, height: height))
        let image = renderer.image { ctx in
            if border {
                // 黑色描边（八方向偏移绘制，偏移量随字号缩放）
                let strokeAttrs: [NSAttributedString.Key: Any] = [
                    .font: font,
                    .foregroundColor: UIColor.black.withAlphaComponent(CGFloat(opacity)),
                ]
                let o = strokeOffset
                for (dx, dy) in [(-o, 0), (o, 0), (0, -o), (0, o), (-o * 0.75, -o * 0.75), (o * 0.75, o * 0.75), (-o * 0.75, o * 0.75), (o * 0.75, -o * 0.75)] {
                    (text as NSString).draw(
                        at: CGPoint(x: pad + dx, y: pad + dy),
                        withAttributes: strokeAttrs)
                }
            }
            (text as NSString).draw(at: CGPoint(x: pad, y: pad), withAttributes: attrs)
        }

        guard let data = image.pngData() else { return nil }
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("text_wm_\(Int(Date().timeIntervalSince1970 * 1000)).png")
        do {
            try data.write(to: url)
        } catch {
            return nil
        }
        return TextPngResult(file: url, width: width, height: height)
    }

    /// 九宫格位置 → 画布上的 overlay 坐标
    static func calcOverlayPosition(_ position: String, _ pngW: Int, _ pngH: Int,
                                    _ canvasW: Int, _ canvasH: Int) -> (Int, Int) {
        let pad = 12
        switch position {
        case "top-left": return (pad, pad)
        case "top-center": return ((canvasW - pngW) / 2, pad)
        case "top-right": return (canvasW - pngW - pad, pad)
        case "center-left": return (pad, (canvasH - pngH) / 2)
        case "center": return ((canvasW - pngW) / 2, (canvasH - pngH) / 2)
        case "center-right": return (canvasW - pngW - pad, (canvasH - pngH) / 2)
        case "bottom-left": return (pad, canvasH - pngH - pad)
        case "bottom-center": return ((canvasW - pngW) / 2, canvasH - pngH - pad)
        default: return (canvasW - pngW - pad, canvasH - pngH - pad) // bottom-right
        }
    }

    private static func parseColor(_ s: String) -> UIColor {
        switch s.lowercased() {
        case "white": return .white
        case "black": return .black
        case "red": return .red
        case "yellow": return .yellow
        case "green": return .green
        case "blue": return .blue
        default:
            var hex = s.replacingOccurrences(of: "#", with: "")
            if hex.count == 6, let v = UInt64(hex, radix: 16) {
                return UIColor(red: CGFloat((v >> 16) & 0xFF) / 255,
                               green: CGFloat((v >> 8) & 0xFF) / 255,
                               blue: CGFloat(v & 0xFF) / 255, alpha: 1)
            }
            hex = ""
            return .white
        }
    }
}
